### PR TITLE
ARROW-13617: [C++] Make Decimal representations consistent

### PR DESCRIPTION
**Pull request by pitrou**
Factor out some basics of decimal representation and implementation in a generic base class.

### CODE REVIEW

**Review by lidavidm — APPROVED** [review]
LGTM, thanks for doing this.

**Comment by lidavidm** [comment]
@ursabot please benchmark lang=C++

**Comment by lidavidm** [comment]
Hmm, hopefully Travis also gets around to building this PR so that we can run the tests on big-endian.

**Comment by pitrou** [comment]
Unfortunately, Travis never runs on my PRs. I have no idea why.

**Comment by lidavidm** [comment]
Indeed, this PR isn't in the Travis queue, and I also can't manually trigger a build for a fork from the UI, which is unfortunate.

**Comment by pitrou** [comment]
I ran the s390x tests locally under emulation and they seemed to work (I had to disable Gandiva, though).

File: cpp/src/arrow/util/basic_decimal.h [hunk 1]
```diff
@@ -37,50 +38,116 @@ enum class DecimalStatus {
   kRescaleDataLoss,
 };
 
+template <typename Derived, int BIT_WIDTH, int NWORDS = BIT_WIDTH / 64>
```
Comment (cyb70289) [inline cpp/src/arrow/util/basic_decimal.h:41]: Looks `Derived` is not used?

File: cpp/src/arrow/util/basic_decimal.cc [hunk 2]
```diff
@@ -1391,4 +1347,8 @@ BasicDecimal256 operator/(const BasicDecimal256& left, const BasicDecimal256& ri
   return result;
 }
 
+// Explicitly instantiate template base class, for DLL linking on Windows
```
Comment (cyb70289) [inline cpp/src/arrow/util/basic_decimal.cc:1350]: Does this approach avoid putting all template class implementation code inside the header?

**Comment by cyb70289** [comment]
> I ran the s390x tests locally under emulation and they seemed to work (I had to disable Gandiva, though).

IBM provides free s390x cloud instances. I tried, it works. But it expires in half year and you have to apply for a new one.
https://developer.ibm.com/gettingstarted/ibm-linuxone/

File: cpp/src/arrow/util/basic_decimal.h (hunk 1 above)
Comment (pitrou) [reply cpp/src/arrow/util/basic_decimal.h:41]: Indeed.

File: cpp/src/arrow/util/basic_decimal.cc (hunk 2 above)
Comment (pitrou) [reply cpp/src/arrow/util/basic_decimal.cc:1350]: It should, though that's not the goal here :-)

**Pull request closed without merging**